certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,556 @@
+}
+
+
+// STATUS - verified
+// The `_scheduledExecutions` array only changes it's length
+// when one of the schedule functions is called.
+// https://prover.certora.com/output/40577/2ccdf706be7e404e9ada9ecddec30ab8/?anonymousKey=d9e3be8168c4a8290ba7d5480acb9dd4afd7ffb4
+rule scheduledExecutionsCanBeChangedOnlyByScheduleFunctions(env e, method f) {
+    uint256 lengthBefore = getSchedExeLength();
+
+    require(lengthBefore < max_uint256);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    uint256 lengthAfter = getSchedExeLength();
+
+    // If the number of scheduled executions changed, it was increased by one.
+    assert lengthBefore != lengthAfter =>
+        f.selector == sig:scheduleRootChange(address, address[]).selector ||
+        f.selector == sig:scheduleDelayChange(bytes32, uint256, address[]).selector ||
+        f.selector == sig:scheduleGrantDelayChange(bytes32, uint256, address[]).selector ||
+        f.selector == sig:scheduleRevokeDelayChange(bytes32, uint256, address[]).selector ||
+        f.selector == sig:schedule(address, bytes, address[]).selector ||
+        f.selector == sig:scheduleGrantPermission(bytes32, address, address, address[]).selector ||
+        f.selector == sig:scheduleRevokePermission(bytes32, address, address, address[]).selector,
+        "Scheduled executions modified by non-schedule function.";
+}
+
+// STATUS - verified
+// The grant delay of and action (stored in `_grantDelays`) is changed only
+// by calling `setGrantDelay` and the caller must be the `_executionHelper`.
+// https://prover.certora.com/output/40577/4fc440c4a1654fdda243cc1a30151573/?anonymousKey=2f909d3f7a3fe4ed878c24c6e69892b127296685
+rule grantDelaysCanBeChangedOnlyBySetGrantDelay(env e, method f) {
+    bytes32 actionId;
+    uint256 delayBefore = getActionIdGrantDelay(actionId);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    uint256 delayAfter = getActionIdGrantDelay(actionId);
+
+    // If the number of scheduled executions changed, it was increased by one.
+    assert delayBefore != delayAfter =>
+        f.selector == sig:setGrantDelay(bytes32, uint256).selector,
+        "_grantDelays modified by function other than setGrantDelay.";
+    assert delayAfter != delayBefore =>
+            e.msg.sender == getTimelockExecutionHelper(),
+        "Delay changed by other entity than execution helper.";
+}
+
+
+// STATUS - verified
+// https://prover.certora.com/output/40577/457f4ae249cf4f6c96d09aa709578277/?anonymousKey=5c0926936c949b5e66bccc3c1969fbaa236c2e94
+rule revokeDelaysCanBeChangedOnlyBySetRevokeDelay(env e, method f) {
+    bytes32 actionId;
+    uint256 delayBefore = getActionIdRevokeDelay(actionId);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    uint256 delayAfter = getActionIdRevokeDelay(actionId);
+
+    // If the number of scheduled executions changed, it was increased by one.
```

Comment:
I don't see where it's implemented in this rule

Reply:
This is checked in other rule `scheduledExecutionsArrayIsNeverShortened` so I deleted the comment.